executor/aggregate.go

```diff
@@ -45,39 +103,493 @@ type HashAggExec struct {
+	e.inputCh = make(chan *HashAggInput, e.partialConcurrency)
+	e.finishCh = make(chan struct{}, 1)
+
+	e.partialOutputChs = make([]chan *HashAggIntermData, e.finalConcurrency)
+	for i := range e.partialOutputChs {
+		e.partialOutputChs[i] = make(chan *HashAggIntermData, e.partialConcurrency)
+	}
+	e.partialWorkers = make([]HashAggPartialWorker, e.partialConcurrency)
+	e.finalWorkers = make([]HashAggFinalWorker, e.finalConcurrency)
+	// Init partial workers.
+	for i := 0; i < e.partialConcurrency; i++ {
+		groupSets := make([]*mvmap.MVMap, e.finalConcurrency)
+		aggCtxsMappers := make([]aggCtxsMapper, e.finalConcurrency)
+		w := HashAggPartialWorker{
+			aggWorker: aggWorker{
+				finishCh:     e.finishCh,
+				aggFuncs:     e.AggFuncs,
+				groupByItems: e.GroupByItems,
+				maxChunkSize: e.maxChunkSize,
+				WaitGroup:    e.partialWorkerWaitGroup,
+				groupVals:    make([][]byte, 0, 8),
+			},
+			inputCh:               make(chan *chunk.Chunk, 1),
+			aggCtxsMapperHolderCh: make([]chan aggCtxsMapper, e.finalConcurrency),
+			outputChs:             e.partialOutputChs,
+			giveBackCh:            e.inputCh,
+			globalOutputCh:        e.finalOutputCh,
+			groupSets:             groupSets,
+			aggCtxsMaps:           aggCtxsMappers,
+		}
+		e.inputCh <- &HashAggInput{
+			chk:        e.children[0].newChunk(),
+			giveBackCh: w.inputCh,
+		}
+		for j := 0; j < e.finalConcurrency; j++ {
+			w.aggCtxsMapperHolderCh[j] = make(chan aggCtxsMapper, 1)
+			w.aggCtxsMapperHolderCh[j] <- make(aggCtxsMapper, 0)
+		}
+		e.partialWorkers[i] = w
+	}
+
+	// Init final workers.
+	finalAggFuncs := e.newFinalAggFuncs()
+	for i := 0; i < e.finalConcurrency; i++ {
+		e.finalWorkers[i] = HashAggFinalWorker{
+			aggWorker: aggWorker{
+				finishCh:     e.finishCh,
+				aggFuncs:     finalAggFuncs,
+				maxChunkSize: e.maxChunkSize,
+				WaitGroup:    e.finalWorkerWaitGroup,
+				groupVals:    make([][]byte, 0, 8),
+			},
+			aggCtxsMap:          make(aggCtxsMapper, 0),
+			groupMap:            mvmap.NewMVMap(),
+			inputCh:             e.partialOutputChs[i],
+			outputCh:            e.finalOutputCh,
+			finalResultHolderCh: make(chan *chunk.Chunk, 1),
+			rowBuffer:           make([]types.Datum, 0, e.Schema().Len()),
+			mutableRow:          chunk.MutRowFromTypes(e.retTypes()),
+		}
+		e.finalWorkers[i].finalResultHolderCh <- e.newChunk()
+	}
+}
+
+func (e *HashAggExec) newFinalAggFuncs() (newAggFuncs []aggregation.Aggregation) {
+	newAggFuncs = make([]aggregation.Aggregation, 0, len(e.AggFuncs))
+	idx := 0
+	for _, af := range e.AggFuncs {
+		var aggFunc aggregation.Aggregation
+		idx, aggFunc = af.GetFinalAggFunc(idx)
+		newAggFuncs = append(newAggFuncs, aggFunc)
+	}
+	return newAggFuncs
+}
+
+// HashAggPartialWorker gets and handles origin data or partial data from inputCh,
+// then shuffle the intermediate results to corresponded final workers.
+func (w *HashAggPartialWorker) run(ctx sessionctx.Context, finalConcurrency int) {
```

Comment:
We can add a `WaitGroup` parameter to this function and remove the `WaitGroup` stored in the `HashAggPartialWorker` to reduce the code complexity.